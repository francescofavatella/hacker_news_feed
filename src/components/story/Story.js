import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { longText, shortText } from './StoryUtils';
import styles from './Story.module.css';

const WithLink = props =>
  props.expanded ? (
    <a
      href={props.url}
      className={props.className + ' ' + styles.cursor}
      target={'_blank'}
      onClick={e => e.stopPropagation()}
    >
      {props.children}
    </a>
  ) : (
    <div className={props.className}>{props.children}</div>
  );

export class Story extends Component {
  static propTypes = {
    score: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    storyUrl: PropTypes.string.isRequired,
    upvotingUrl: PropTypes.string.isRequired,
    expanded: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    score: -1,
    title: '',
    author: '',
    text: '',
    upvotingUrl: '',
    storyUrl: '',
    expanded: false,
  };

  createMarkup = str => {
    return { __html: str };
  };
  render() {
    const {
      score,
      title,
      author,
      text,
      upvotingUrl,
      storyUrl,
      expanded,
    } = this.props;

    const storyText = expanded ? text : shortText(text);
    const storyTitle = expanded ? longText(title) : title;

    return (
      <div className={styles.container}>
        <div className={styles.leftColumn}>
          <WithLink expanded={expanded} url={upvotingUrl} className={''}>
            <div className={styles.arrowUp} />
            <div className={styles.score}>{score}</div>
          </WithLink>
        </div>
        <div className={styles.rigthColumn}>
          <WithLink expanded={expanded} url={storyUrl} className={styles.title}>
            {storyTitle}
          </WithLink>
          <div className={styles.author}>{'By ' + author}</div>
          <div
            className={styles.text}
            dangerouslySetInnerHTML={this.createMarkup(storyText)}
          />
        </div>
      </div>
    );
  }
}

export default Story;
