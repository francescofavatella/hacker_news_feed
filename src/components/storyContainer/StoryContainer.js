import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Story from '../story/Story';
import styles from './StoryContainer.module.css';
import storyLoader from './StoryLoader';

export const StoryLoading = ({ storyId }) => <li key={storyId}>loading...</li>;
export const StoryFailed = ({ storyId }) => (
  <li key={storyId}>Error loading content</li>
);

export class StoryContainer extends Component {
  static propTypes = {
    storyId: PropTypes.number.isRequired,
  };
  static defaultProps = {
    storyId: -1,
  };

  constructor() {
    super();
    this.state = {
      storyProps: undefined,
      loading: true,
      expanded: false,
      failed: false,
    };
  }
  componentDidMount() {
    this.fetchStory();
  }

  fetchStory = async () => {
    try {
      const newStory = await storyLoader(this.props.storyId);
      this.setState((state, props) => ({
        storyProps: Object.assign({}, newStory),
        loading: false,
      }));
    } catch (err) {
      this.setState((state, props) => ({
        failed: true,
        loading: false,
      }));
    }
  };

  onClickStory = () =>
    this.setState((state, props) => ({
      expanded: !this.state.expanded,
    }));

  render() {
    const { storyId } = this.props;
    if (this.state.failed) {
      return <StoryFailed storyId={storyId} />;
    }
    if (this.state.loading) {
      return <StoryLoading storyId={storyId} />;
    }
    const { expanded } = this.state;

    const containerClassName = expanded
      ? styles.storyContainerLarge
      : styles.storyContainerSmall;

    return (
      <li
        onClick={this.onClickStory}
        className={styles.storyContainer + ' ' + containerClassName}
      >
        <Story {...this.state.storyProps} expanded={expanded} />
      </li>
    );
  }
}
export default StoryContainer;
