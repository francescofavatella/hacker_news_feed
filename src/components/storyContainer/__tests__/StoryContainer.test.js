import React from 'react';
import TestRenderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import StoryContainer, { StoryLoading, StoryFailed } from '../StoryContainer';
import Story from '../../story/Story';
import storyLoader from '../StoryLoader';

jest.mock('../StoryLoader');

describe('StoryContainer', () => {
  let resultMock;
  beforeEach(() => {
    resultMock = {
      score: 11,
      title: 'Title_mock',
      author: 'Author_mock',
      text: 'Text_mock',
      upvotingUrl: 'upvotingUrl_mock',
      storyUrl: 'storyUrl_mock',
      expanded: false,
    };
  });
  afterEach(() => {
    storyLoader.mockClear();
  });

  describe('with no props', () => {
    const component = <StoryContainer />;

    it('renders without crashing', () => {
      shallow(component);
    });

    it('matches snapshot', () => {
      const tree = TestRenderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('loading data', async () => {
    it('should show Story if success', done => {
      expect.assertions(7);

      const fetch = async () => Promise.resolve(resultMock);
      storyLoader.mockImplementation(fetch);

      const wrapper = shallow(<StoryContainer storyId={1} />);
      expect(wrapper.find(StoryLoading).length).toEqual(1);
      setTimeout(() => {
        wrapper.update();
        expect(storyLoader.mock.calls.length).toEqual(1);
        expect(wrapper.instance().state.loading).toEqual(false);
        expect(wrapper.instance().state.storyProps).toEqual(resultMock);
        const createdStory = wrapper.find(Story);
        expect(createdStory.length).toEqual(1);
        expect(createdStory.props()).toEqual(resultMock);
        expect(wrapper).toMatchSnapshot();
        done();
      });
    });

    it('should show StoryFailed if error', async done => {
      expect.assertions(5);

      const fetch = async () => Promise.reject(new Error('error'));
      storyLoader.mockImplementation(fetch);

      const wrapper = shallow(<StoryContainer storyId={1} />);
      expect(wrapper.find(StoryLoading).length).toEqual(1);
      setTimeout(() => {
        wrapper.update();

        expect(storyLoader.mock.calls.length).toEqual(1);
        expect(wrapper.instance().state.loading).toEqual(false);
        expect(wrapper.find(StoryFailed).length).toEqual(1);
        expect(wrapper).toMatchSnapshot();
        done();
      });
    });
  });

  describe('on click event', () => {
    it('should expand the story if it is closed', () => {
      const wrapper = shallow(<StoryContainer storyId={1} />);
      wrapper.setState({
        storyProps: resultMock,
        loading: false,
      });

      expect(wrapper.state().expanded).toEqual(false);
      expect(wrapper.find('li').hasClass('storyContainerSmall')).toEqual(true);

      wrapper.find('li').simulate('click');

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().expanded).toEqual(true);
      expect(wrapper.find('li').hasClass('storyContainerLarge')).toEqual(true);
    });

    it('should close the story if it is expanded', () => {
      const wrapper = shallow(<StoryContainer storyId={1} />);
      wrapper.setState({
        storyProps: resultMock,
        loading: false,
        expanded: true,
      });

      expect(wrapper.state().expanded).toEqual(true);
      expect(wrapper.find('li').hasClass('storyContainerLarge')).toEqual(true);

      wrapper.find('li').simulate('click');

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().expanded).toEqual(false);
      expect(wrapper.find('li').hasClass('storyContainerSmall')).toEqual(true);
    });
  });
});
