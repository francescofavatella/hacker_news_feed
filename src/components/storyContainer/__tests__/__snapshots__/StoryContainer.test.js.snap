// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryContainer loading data should show Story if success 1`] = `
<li
  className="storyContainer storyContainerSmall"
  onClick={[Function]}
>
  <Story
    author="Author_mock"
    expanded={false}
    score={11}
    storyUrl="storyUrl_mock"
    text="Text_mock"
    title="Title_mock"
    upvotingUrl="upvotingUrl_mock"
  />
</li>
`;

exports[`StoryContainer loading data should show StoryFailed if error 1`] = `
<StoryFailed
  storyId={1}
/>
`;

exports[`StoryContainer on click event should close the story if it is expanded 1`] = `
<li
  className="storyContainer storyContainerSmall"
  onClick={[Function]}
>
  <Story
    author="Author_mock"
    expanded={false}
    score={11}
    storyUrl="storyUrl_mock"
    text="Text_mock"
    title="Title_mock"
    upvotingUrl="upvotingUrl_mock"
  />
</li>
`;

exports[`StoryContainer on click event should expand the story if it is closed 1`] = `
<li
  className="storyContainer storyContainerLarge"
  onClick={[Function]}
>
  <Story
    author="Author_mock"
    expanded={true}
    score={11}
    storyUrl="storyUrl_mock"
    text="Text_mock"
    title="Title_mock"
    upvotingUrl="upvotingUrl_mock"
  />
</li>
`;

exports[`StoryContainer with no props matches snapshot 1`] = `
<li>
  loading...
</li>
`;
